compiler/hash-utils/src/graph/mod.rs

```diff
@@ -0,0 +1,67 @@
+//! This module contains graph algorithms and utilities that are used by the
+//! compiler.
+
+pub mod dominators;
+pub mod visit;
+
+use index_vec::Idx;
+
+/// This trait can be used to represent a generic directed
+/// graph. The index of each graph node used a [`index_vec::Idx`].
+pub trait DirectedGraph {
```

Comment:
Surely one needs more info to call this a directed graph?

Reply:
Why is DirectedGraph split into WithPredecessors and WithSuccessors?

Reply:
Yeah it could be just combined into `DirectedGraph`

Reply:
Also, what additional information is needed to call it a `DirectedGraph`?

Reply:
Well a directed graph is a set of nodes N and for each ordered pair of nodes (a, b) in NxN, an optional edge.

For example,

```rust
pub trait DirectedGraph {
   type Node;
   type Edge;
   type SuccessorsIter;
   type PredecessorsIter;

   // Optional method to get an edge between two nodes if it exists
   // By default, it can be implemented as `self.successors(a).find_map(|&x| x == b)`
   fn edge(&self, a: Self::Node, b: Self::Node) -> Optional<Self::Edge>;

   fn successors(&self, a: Self::Node) -> Self::SuccessorsIter;
   fn predecessors(&self, a: Self::Node) -> Self::PredecessorsIter;
}
```

Reply:
There isn’t much need to access this information, only successor and predecessor queries are required

Reply:
Also it's probably better if `Node` is a generic parameter rather than an assoc type

Reply:
Any particular reason?